rbm_softmax.py

```diff
@@ -0,0 +1,504 @@
+        Parameters
+        ----------
+        v_pos : array-like, shape (n_samples, n_features)
+            The data to use for training.
+        """
+        h_pos = self._mean_hiddens(v_pos)
+        v_neg = self._sample_visibles(self.h_samples_)
+        h_neg = self._mean_hiddens(v_neg)
+
+        lr = float(self.learning_rate) / v_pos.shape[0]
+        update = safe_sparse_dot(v_pos.T, h_pos, dense_output=True).T
+        update -= np.dot(h_neg.T, v_neg)
+        self.components_ += lr * update
+        self.intercept_hidden_ += lr * (h_pos.sum(axis=0) - h_neg.sum(axis=0))
+        self.intercept_visible_ += lr * (np.asarray(
+                                         v_pos.sum(axis=0)).squeeze() -
+                                         v_neg.sum(axis=0))
+
+        h_neg[self.rng_.uniform(size=h_neg.shape) < h_neg] = 1.0  # sample binomial
+        self.h_samples_ = np.floor(h_neg, h_neg)
+
+    def corrupt(self, v):
+        # Randomly corrupt one feature in each sample in v.
+        ind = (np.arange(v.shape[0]),
+               self.rng_.randint(0, v.shape[1], v.shape[0]))
+        if issparse(v):
+            data = -2 * v[ind] + 1
+            v_ = v + sp.csr_matrix((data.A.ravel(), ind), shape=v.shape)
+        else:
+            v_ = v.copy()
+            v_[ind] = 1 - v_[ind]
+        return v_, None
+
+    def uncorrupt(self, visibles, state):
+        pass
+
+    @common.timeit
+    def score_samples(self, X):
+        """Compute the pseudo-likelihood of X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} shape (n_samples, n_features)
+            Values of the visible layer. Must be all-boolean (not checked).
+
+        Returns
+        -------
+        pseudo_likelihood : array-like, shape (n_samples,)
+            Value of the pseudo-likelihood (proxy for likelihood).
+
+        Notes
+        -----
+        This method is not deterministic: it computes a quantity called the
+        free energy on X, then on a randomly corrupted version of X, and
+        returns the log of the logistic function of the difference.
+        """
+        check_is_fitted(self, "components_")
+
+        v = check_array(X, accept_sparse='csr')
+        fe = self._free_energy(v)
+
+        v_, state = self.corrupt(v)
+        # TODO: If I wanted to be really fancy here, I would do one of those "with..." things.
+        fe_corrupted = self._free_energy(v)
+        self.uncorrupt(v, state)
+        return fe.mean(), fe_corrupted.mean()
+
+        # TODO: I don't have a great intuition about this. Why multiply by n_features?
+        # The overfitting section of "Practical Guide" just says to compare the
+        # average free energy of train and validation and to compare them. Any reason
+        # not to just do that here as well? Seems much simpler to interpret. Maybe
+        # it's because we're supposed to be dealing with smaller deltas here?
+        #return v.shape[1] * log_logistic(fe_corrupted - fe)
```

Comment:
@falsifian the commented line above is what was used in the original implementation. Any idea why it would be used rather than the seemingly-more-straightforward thing  I did? I found this number confusing to interpret ("if the energy of the training data goes down relative to the corrupted version, then the sigmoid goes to -1, so the log goes to 0 from the left, so...")